### PR TITLE
NO-TICKET - convert Web3Auth JS callbacks to coroutines

### PR DESCRIPTION
**Pull request by gslevinkov**
## Description 
- convert Web3Auth JS callbacks to coroutines

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/auth/web3authsdk/Web3AuthApiClient.kt [hunk 1]
```diff
@@ -142,48 +146,37 @@ class Web3AuthApiClient(
...
-                ?.let { handler?.handleApiError(it) }
-                ?: handleMapperError(error)
-            handler = null
-        }
-
-        private fun handleMapperError(originalResponse: String) {
-            val error = Web3AuthSdkInternalError("Web3Auth SDK method result parsing failed: $originalResponse")
-            Timber.e(error)
-            handler?.handleInternalError(error)
+            runCatching<Throwable> { mapper.fromNetworkError(error) }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/web3authsdk/Web3AuthApiClient.kt:177]: uuuh, duck-taped, but still working idea

File: app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt [hunk 2]
```diff
@@ -1,30 +1,39 @@
...
 
-    fun fromNetworkSignIn(responseJson: String): Web3AuthSignInResponse? {
-        return kotlin.runCatching { gson.fromJsonReified<Web3AuthSignInResponse>(responseJson) }
-            .onFailure { Timber.i(it) }
-            .getOrNull()
+    fun fromNetworkSignIn(responseJson: String): Result<Web3AuthSignInResponse> = try {
+        Result.success(gson.fromJsonReified<Web3AuthSignInResponse>(responseJson)!!)
+    } catch (mappingError: Throwable) {
+        Timber.i(mappingError)
+        Timber.i(responseJson)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt:26]: it would be wise to log broken JSON what you think?

Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt:26]: but can be logged in same line

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/auth/web3authsdk/Web3AuthApiClient.kt [hunk 3]
```diff
@@ -45,72 +44,77 @@ class Web3AuthApiClient(
...
-                type = "signup",
-                jsMethodCall = "triggerSilentSignup('$socialShare')"
-            ),
-            null
-        )
+    override suspend fun triggerSilentSignUp(socialShare: String): Web3AuthSignUpResponse {
+        return suspendCancellableCoroutine {
+            this.continuation = it
+
+            onboardingWebView.evaluateJavascript(
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/auth/web3authsdk/Web3AuthApiClient.kt:51]: You can move it to func with 2 args evaluateJavascript(type, methodCall)

File: app/src/main/java/org/p2p/wallet/auth/web3authsdk/Web3AuthApiClient.kt [hunk 4]
```diff
@@ -45,72 +44,77 @@ class Web3AuthApiClient(
...
-        val deviceShareAsJsObject = gson.toJson(deviceShare)
-        onboardingWebView.evaluateJavascript(
-            generateFacade(
-                type = "signin",
-                jsMethodCall = "triggerSignInNoCustom('$socialShare', $deviceShareAsJsObject)"
-            ),
-            null
-        )
+    ): Web3AuthSignInResponse {
+        return suspendCancellableCoroutine {
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/auth/web3authsdk/Web3AuthApiClient.kt:83]: and move this part of setting continuation to function too

File: app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt [hunk 5]
```diff
@@ -1,30 +1,39 @@
...
+        Result.failure(Web3AuthSdkInternalError(INTERNAL_ERROR_MESSAGE, mappingError))
     }
 
-    fun fromNetworkError(errorResponseJson: String): Web3AuthErrorResponse? {
-        return kotlin.runCatching { gson.fromJsonReified<Web3AuthErrorResponse>(errorResponseJson) }
-            .onFailure { Timber.i(it) }
-            .getOrNull()
-            ?.let { createFilledResponseWithErrorType(it) }
+    fun fromNetworkError(errorResponseJson: String): Web3AuthErrorResponse = try {
+        val web3AuthError = gson.fromJsonReified<Web3AuthErrorResponse>(errorResponseJson)!!
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt:31]: can we avoid !! here ?

**Review by davran1001 — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt (hunk 2 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt:26]: im not sure how :(

File: app/src/main/java/org/p2p/wallet/auth/web3authsdk/Web3AuthApiClient.kt (hunk 3 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/auth/web3authsdk/Web3AuthApiClient.kt:51]: i would rather not to touch it - the code was written by Kostya, and it's very dangerous

File: app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt (hunk 5 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/auth/web3authsdk/mapper/Web3AuthClientMapper.kt:31]: no, I tried, but i haven't found a way

**Pull request merged**